runtime/nvqir/CircuitSimulator.h

```diff
@@ -146,16 +146,16 @@ class CircuitSimulator {
     std::vector<std::size_t> qubitSupport;
     std::vector<std::function<void(bool)>> basisChange;
     op.for_each_pauli([&](cudaq::pauli type, std::size_t qubitIdx) {
+      auto qId = qubitIds[qubitIdx];
       if (type != cudaq::pauli::I)
-        qubitSupport.push_back(qubitIds[qubitIdx]);
+        qubitSupport.push_back(qId);
 
       if (type == cudaq::pauli::Y)
-        basisChange.emplace_back([&, qubitIdx](bool reverse) {
-          rx(!reverse ? M_PI_2 : -M_PI_2, qubitIds[qubitIdx]);
+        basisChange.emplace_back([this, qId](bool reverse) {
+          rx(!reverse ? M_PI_2 : -M_PI_2, qId);
```

Comment:
2 points to the person that explains to me -- without falling back on a hypothetical compiler bug explanation -- why the Python changes in this PR require this change. And only in gcc11/x86 builds when paired side-by-side with the installer....on some (but not all) operating systems. It worked fine in gcc11/ARM builds (all operating systems) in the "Side-by-side with installer" test, too, so it really was only gc11+x86+side-by-side-with-installer-on-some-operating-systems.

All other standard Docker tests and wheel tests pass with gcc12, clang16, and llvm bootstrapped builds (both before and after this change). 

In any case, if we need to make this change, I think we have similar code in `runtime/nvqir/cutensornet/simulator_mps_register.cpp`, so we should probably put it in that file as well.

I'm open to other ideas if people have them.

Reply:
I didn't confirm this, but it looks a C++ coding bug.

1. There is the `for_each_pauli` iterator that applies a lambda to each element.
1. Each time, a new instance of the lambda is created with an argument `qubitIdx`.
1. When the type is X or Y, a second lambda is instantiated and pushed on the `basisChange` queue.
1. The second lambda captures a reference to `qubitIdx` from the outer lambda.
1. When the `for_each_pauli` ends, all the outer lambdas are no longer live, so the compiler could reclaim them.
1. `basisChange` is still live, still contains the inner lambdas, and the inner lambdas still hold references to the `qubitIdx` arguments that were in the now defunct and possibly reclaimed outer lambdas.

Passing the value of the element by-value to the inner lambda(s) looks like the correct fix then.

Reply:
Trying again. The inner lambdas are causing a reference capture of the reference qubitIds argument. It seems like this should work fine (and it seems to with most of the compilers Ben has tested). But apparently, it is also known to potentially tickle some bugs. Here is a discussion: https://stackoverflow.com/questions/21443023/capturing-a-reference-by-reference-in-a-c11-lambda